src/Workspaces/Core/Portable/NamingStyles/Serialization/SymbolSpecification.cs

```diff
@@ -228,9 +229,35 @@ internal static SymbolKindOrTypeKind AddTypeKindFromXElement(XElement typeKindEl
             {
                 return new SymbolKindOrTypeKind((TypeKind)Enum.Parse(typeof(TypeKind), typeKindElement.Value));
             }
+
+            public override bool Equals(object obj)
+            {
+                return Equals(obj as SymbolKindOrTypeKind);
+            }
+
+            public bool Equals(SymbolKindOrTypeKind other)
+            {
+                return other != null &&
+                       this.SymbolKind == other.SymbolKind &&
+                       this.TypeKind == other.TypeKind;
+            }
+
+            public override int GetHashCode()
+            {
+                if (this.SymbolKind.HasValue && this.TypeKind.HasValue)
```

Comment:
How is it possible to have both? There's constructors assigning one or the other. Is this mutated outside the class?